feat: manual stack assets

shenlong-tanwen opened the pull request:
Initial support for #2479 

#### Changes made in the PR:

- Initial support for Stacked assets where assets can be manually added or removed from a stack
- Option to change the main asset of a stack
- New column `stackParentId` added to `AssetEntity` which would hold the id of the stack parent for an asset
- View and navigate to stacked children in Mobile app asset-viewer
- Deleting a stack parent would pick an asset from the children to be the new parent. Trashing would move the stack parent to the trash and children will still be hidden from the timeline, deleting the parent from trash would pick a new parent. Restoring would restore the entire stack.
- Stacking an asset which has children to a new parent will merge the children to the new parent as well
- Stacking is only handled in the main timeline. All other views such as Favorites, Archive, Albums etc always shows individual assets.
- Multiple assets can be selected from the main timeline and stacked together
- Sharing is asset specific, i.e, sharing a parent does not imply that all the children are shared / added to an album as well. Children has to be explicitly shared / added to an album.
- Downloading is asset specific as well. i.e, Stacked children has to be downloaded individually. Downloading the entire stack can be added in a future PR.
- Asset selection page is refactored in mobile to make it possible to deselect assets as well.
- Asset selection order is now maintained using a `LinkedHashSet` instead of the previous `HashSet`. This gives a more consistent behavior to the user.


[stack-mobile-opt.webm](https://github.com/immich-app/immich/assets/139912620/aa9f0701-8b02-4d96-8277-c9b75a39777f)


#### TODO:
- [ ] Stack / Unstack from Web
- [ ] View stacked children in web
- [x] Stacked indicator to thumbnail
- [ ] Ensure stacked assets are searchable
- [x] Handle archive / favorite / delete of stacked assets
- [x] Stacked assets and sharing
- [x] Stacked assets and albums
- [ ] Stacked assets and downloading
- [x] Add tests

shenlong-tanwen committed:
feat(server): add stackParentId to asset entity

shenlong-tanwen committed:
feat(mobile): stackParentId to Asset

shenlong-tanwen committed:
feat(mobile): stack / unstack assets

shenlong-tanwen committed:
chore: rebase main

shenlong-tanwen committed:
refactor(mobile): asset_selection_page

shenlong-tanwen committed:
fix(mobile): block stacked assets from getting added to albums

shenlong-tanwen committed:
chore: rebase main

shenlong-tanwen committed:
refactor(mobile): gallery viewer

shenlong-tanwen committed:
chore(server): format

shenlong-tanwen committed:
server: replace stack parent on parent trash / deletion

shenlong-tanwen committed:
server: merge stacks

shenlong-tanwen committed:
mobile: stack row styling

shenlong-tanwen committed:
fix(server): show assets as unstacked in trash

shenlong-tanwen committed:
chore: rebase main

shenlong-tanwen committed:
mobile: handle stack child deletion and archive

shenlong-tanwen committed:
feat: change stack parent

shenlong-tanwen committed:
mobile: stack assets using multiselection

shenlong-tanwen committed:
mobile: preserve selection order

shenlong-tanwen committed:
mobile: unstack option

shenlong-tanwen committed:
server: do not update parent on trash

shenlong-tanwen committed:
fix(mobile): lazy load stack thumbs

shenlong-tanwen committed:
show stacks in trash page

alextran1502 committed:
merge main

shenlong-tanwen committed:
stack indicator

alextran1502 committed:
fix response dto

shenlong-tanwen committed:
chore: pull main

shenlong-tanwen committed:
chore: pull main

shenlong-tanwen committed:
mobile: fetch stack for only assets with children

shenlong-tanwen committed:
test: fix pr checks

shenlong-tanwen committed:
mobile: only build stack list when asset has children

shenlong-tanwen committed:
mobile: add translation texts

shenlong-tanwen committed:
server: update stack parent only if stack is updated

shenlong-tanwen committed:
test: stack unit tests

shenlong-tanwen committed:
test: e2e test for stacks

shenlong-tanwen committed:
server: do not hide stacked assets from timeline yet

shenlong-tanwen committed:
chore(mobile): show border only for selected child

shenlong-tanwen committed:
mobile: show stack indicator only for image assets

shenlong-tanwen committed:
chore: pull main

shenlong-tanwen committed:
feat(server): send stack updates through ws

jrasm91 commented on server/src/domain/asset/response-dto/asset-response.dto.ts:
```diff
@@ -42,9 +42,13 @@ export class AssetResponseDto extends SanitizedAssetResponseDto {
 }
 
-export function mapAsset(entity: AssetEntity, stripMetadata = false): AssetResponseDto {
+export function mapAsset(entity: AssetEntity, stripMetadata = false, withStack = false): AssetResponseDto {
```
I think we should change this to be `{ stripMetadata: boolean, withStack: boolean }` instead of two boolean params, as it is hard to know what each one does in other files that call this function.

shalong-tanwen replied:
Updated.

jrasm91 commented on server/src/immich/api-v1/asset/asset.core.ts:
```diff
@@ -48,6 +48,7 @@ export class AssetCore {
       faces: [],
       sidecarPath: sidecarPath || null,
       isReadOnly: dto.isReadOnly ?? false,
+      stackParentId: null,
```
We should not need this change, and it should not be required if it is. It should default to null at the database level when not provided.

shalong-tanwen replied:
Removed. Thank you.

jrasm91 commented on server/src/immich/api-v1/asset/asset.service.ts:
```diff
@@ -196,7 +196,7 @@ export class AssetService {
     const asset = await this._assetRepository.getById(assetId);
     if (includeMetadata) {
-      const data = mapAsset(asset);
+      const data = mapAsset(asset, false, true);
```
This is hard to read imo, we should pass an options object instead.

shalong-tanwen replied:
Updated to an options object.

jrasm91 commented on server/src/immich/controllers/asset.controller.ts:
```diff
@@ -145,4 +147,16 @@ export class AssetController {
     return this.service.update(authUser, id, dto);
   }
+
+  @Post('stack')
```
These should go before the previous route, which is a wildcard `PUT asset/:id`.

shalong-tanwen replied:
Handled.

jrasm91 commented on server/src/immich/controllers/asset.controller.ts:
```diff
@@ -145,4 +147,16 @@ export class AssetController {
+  @Post('stack')
+  @HttpCode(HttpStatus.OK)
+  updateStack(@AuthUser() authUser: AuthUserDto, @Body() dto: UpdateAssetStackDto): Promise<void> {
+    return this.service.updateStack(authUser, dto);
```
Can this be merged/implemented with just `PUT /asset`. I believe this is bulk asset update endpoint and it seems perfect to handle bulk setting the stack parent on multiple assets.

shalong-tanwen replied:
Removed `POST /asset/stack` asset and updated `PUT /asset` to handle stacking as well. Had to add one more entry `removeParent` to `AssetBulkUpdateDto` other than `stackParentId`  because the mobile app cannot send both undefined and null for `stackParentId` to handle parent removal.

jrasm91 commented on server/src/infra/entities/asset.entity.ts:
```diff
@@ -148,6 +148,16 @@ export class AssetEntity {
   @OneToMany(() => AssetFaceEntity, (assetFace) => assetFace.asset)
   faces!: AssetFaceEntity[];
+
+  @Column({ type: 'varchar', nullable: true })
```
This type isn't actually accurate, since it is a uuid from the relation. Just remove `type: 'varchar'` entirely.

shalong-tanwen replied:
Updated.

jrasm91 commented on server/src/infra/entities/asset.entity.ts:
```diff
@@ -148,6 +148,16 @@ export class AssetEntity {
+  @JoinColumn({ name: 'stackParentId' })
+  stackParent?: AssetEntity | null;
+
+  @OneToMany(() => AssetEntity, (asset) => asset.stackParent, { onDelete: 'SET NULL' })
```
It doesn't look like this `onDelete SET NULL` made it into the migration. Can you also add `onUpdate: CASCADE` as well though. I think these settings might been to be in the `stackParent` relation. It means "when the parent is deleted, set the child's `stackParentId` to `NULL`. The one you have here means "when a child is deleted...", which doesn't make sense. Try moving it an re-generating the migration. Also, since there is no set null on delete you shouldn't be able to execute a parent asset delete in the database, which should continue to work.

shalong-tanwen replied:
True. My bad. Made the required changes now. Thanks a lot.

jrasm91 commented on server/test/e2e/asset.e2e-spec.ts:
```diff
@@ -627,4 +627,167 @@ describe(`${AssetController.name} (e2e)`, () => {
+    });
+  });
+
+  describe('POST /asset/stack/parent', () => {
```
```suggestion
  describe('PUT /asset/stack/parent', () => {
```

shalong-tanwen replied:
Updated 😅

jrasm91 reviewed (commented):
Looks pretty good! Thanks for writing e2e tests too. Just a few initial comments.

shenlong-tanwen committed:
refactor(server): map asset options

shenlong-tanwen committed:
chore(server): move updateAsset to the end in asset controller

shenlong-tanwen committed:
fix(server): stackParentId cascade

shenlong-tanwen committed:
use bulkassetupdate to add / remove children to stack

shenlong-tanwen committed:
test: update server tests

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into feat/stacked-photos

alextran1502 commented:
Hello and thank you for the fantastic feature. I am sure many people will appreciate this, especially those who use burst shots often.

One issue I encounter when testing is that when entering the stacking view (i.e. select 1 asset and tap on the stack), the timeline gets rendered with the archived asset. I think we shouldn't render the archive asset to keep it consistent with the view.

shenlong-tanwen committed:
mobile: exclude archived assets from stack selection

shenlong-tanwen committed:
mobile: selection asset state handling

shenlong-tanwen committed:
Merge branch 'main' into feat/stacked-photos

shenlong-tanwen commented:
> One issue I encounter when testing is that when entering the stacking view (i.e. select 1 asset and tap on the stack), the timeline gets rendered with the archived asset. I think we shouldn't render the archive asset to keep it consistent with the view.

Since we can archive assets after stacking them, I thought we'd also give them the option to select archived assets. But hiding them makes sense as well. I've updated the handling to not show them for stack selection
 
> Another one is that when you select the local-only asset, the stack button is shown. I think in the selection assets, if there is a local-only asset, we won't show the stack button

Bottom bar state handling is updated to handle this. I've refactored it to a separate model. It'll be easier in future to add asset storage specific handling to the bottom bar

shenlong-tanwen committed:
chore: fix pr action
